pkg/collector/corechecks/containerlifecycle/queue.go

```diff
@@ -18,81 +18,98 @@ type queue struct {
 	sync.RWMutex
 }
 
+// newQueue returns a new *queue.
 func newQueue(chunkSize int) *queue {
 	return &queue{
 		chunkSize: chunkSize,
 		data:      []model.EventsPayload{},
 	}
 }
 
-func (q *queue) dump() []model.EventsPayload {
-	q.RLock()
-	defer q.RUnlock()
-
-	return q.data
-}
-
-func (q *queue) reset() {
+// flush returns and resets the queue content. Returns nil if the queue is empty.
+// flush is thread-safe.
+func (q *queue) flush() []model.EventsPayload {
 	q.Lock()
 	defer q.Unlock()
 
+	if q.isEmpty() {
+		return nil
+	}
+
+	data := make([]model.EventsPayload, len(q.data))
+	copy(data, q.data)
+
+	// Reset the data in the queue.
 	q.data = []model.EventsPayload{}
+
+	return data
 }
 
+// add enqueues a new event.
+// add is thread-safe.
 func (q *queue) add(ev event) error {
-	if q.isEmpty() || q.isLastPayloadFull() {
-		payload, err := ev.toPayloadModel()
-		if err != nil {
-			return err
-		}
+	q.Lock()
+	defer q.Unlock()
 
-		q.addPayload(payload)
-		return nil
+	if q.isEmpty() || q.isLastPayloadFull() {
+		return q.addPayload(ev)
 	}
 
-	event, err := ev.toEventModel()
+	return q.addEvent(ev)
+}
+
+// addPayload enqueues the event in a new payload entry in the queue.
+// To be used if the queue is empty or if the last payload entry in the queue is full.
+// addPayload is not thread-safe, the caller must lock the queue.
+func (q *queue) addPayload(ev event) error {
+	payload, err := ev.toPayloadModel()
 	if err != nil {
 		return err
 	}
 
-	q.addEvent(event)
+	q.data = append(q.data, payload)
 
 	return nil
 }
 
-func (q *queue) addPayload(payload model.EventsPayload) {
-	q.Lock()
-	defer q.Unlock()
-
-	q.data = append(q.data, payload)
-}
+// addEvent enqueues the event in last payload entry.
+// To be used if the queue is not empty and the last payload entry is not full.
+// addEvent is not thread-safe, the caller must lock the queue.
+func (q *queue) addEvent(ev event) error {
+	if q.isEmpty() {
+		return errors.New("cannot add event to an empty queue")
```

Comment:
This error should be impossible because of the test done line 54.

Reply:
indeed, it's just a safeguard any future use of this method. (in the old code this method wasn't meant to be called when the queue is empty, but it happened anyway and caused the index out of range panic)

Reply:
Yes. But it turns out it was caused by a bug in the code, the race conditions that you fixed in your locking refactoring.
The question is: in case of bug in the code, do we prefer:
* to fail hard (`panic`, either by letting the code crash further in the code or by explicitly invoking `panic()`) in order to detect the bugs and fix them or
* to do our best to make potential bugs more or less silent with the drawback of reporting potentially incorrect/incomplete data for a while before we notice it.